Add MaterialTheme 3 definition to our WooTheme

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Part of WOOMOB-933

### Description
Currently, we are using both material 2 and material 3 components, but we didn't have an explicit definition of the material 3 theme, which meant that when using material 3 components, we'll be using their default look.
This PR adds a definition of the Material 3 theme, that aims to stay consistent as much as possible with our current theme: defining colors, typography and also shapes.

### Steps to reproduce
From what I saw, the only visible component from material 3 that we use is the `ModalBottomSheet`, so for testing:
- Open Blaze campaign creation screen.
- Edit budget or destination URL parameters.
- Check that the bottom sheet looks as before.

Feel free to explore other screens of the app for a quick sanity check.

### Testing information
Confirm there is no changes to the app UI.

### The tests that have been performed
The above.

### Images/gif
| Before | After |
| --- | --- |
| <img width="1080" height="2400" alt="Screenshot_20250729_104815" src="https://github.com/user-attachments/assets/4b7ad87e-ed6e-4984-afaa-29093ef15f07" /> | <img width="1080" height="2400" alt="Screenshot_20250729_105049" src="https://github.com/user-attachments/assets/269e0f51-969b-4889-b5d0-286e31470c92" /> |
| <img width="1080" height="2400" alt="Screenshot_20250729_104806" src="https://github.com/user-attachments/assets/f5d5c8d1-ea27-46aa-8eb9-c113e563b718" /> | <img width="1080" height="2400" alt="Screenshot_20250729_105101" src="https://github.com/user-attachments/assets/1f797bfa-8884-4bbf-82d4-eeeae43bfeca" /> |
| <img width="1080" height="2400" alt="Screenshot_20250729_104729" src="https://github.com/user-attachments/assets/710675b2-2b29-40a3-83ab-0effa463578d" /> | <img width="1080" height="2400" alt="Screenshot_20250729_105135" src="https://github.com/user-attachments/assets/ef0b3256-7f76-48d5-b479-ecb2e9f92797" /> |
| <img width="1080" height="2400" alt="Screenshot_20250729_104754" src="https://github.com/user-attachments/assets/ae3eb3a1-9644-4275-a2c0-764426b60863" /> | <img width="1080" height="2400" alt="Screenshot_20250729_105739" src="https://github.com/user-attachments/assets/710b8695-e07c-4783-9a1f-5b43f103178c" /> |


- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Use the xml defined colors as the default colors for the Compose theme

hichamboushaba committed:
Introduce Material 3 colors and typography definitions

hichamboushaba committed:
Introduce Material3 theme definition

hichamboushaba committed:
fixup! Introduce Material 3 colors and typography definitions

hichamboushaba committed:
Define default material shapes

hichamboushaba committed:
Remove unneeded overrides in the modal bottom sheet

hichamboushaba committed:
Use surface elevated colors for surface containers

Material3 doesn't use elevations, so this is just a temporary measure to ensure more contrast between containers using Surface and the main screen.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/theme/WooColors.kt:
```diff
@@ -2,63 +2,118 @@ package com.woocommerce.android.ui.compose.theme
+        inverseSurface = colorResource(R.color.color_on_surface),
+        inversePrimary = colorResource(R.color.color_on_primary),
+        surfaceBright = colorResource(R.color.color_surface),
+        surfaceContainer = colorResource(R.color.color_surface_elevated),
+        surfaceContainerHigh = colorResource(R.color.color_surface_elevated),
+        surfaceContainerHighest = colorResource(R.color.color_surface_elevated),
+        surfaceContainerLow = colorResource(R.color.color_surface_elevated),
+        surfaceContainerLowest = colorResource(R.color.color_surface_elevated),
```
While Material2 used the elevation of components to apply an overlay to elevated components, Material 3 seems to have moved from this, and the elevation is not even used as often as before, so to keep a consistent look with before, I'm using the `elevated` surface color for all the `surfaceContainer` variations for now, IMO this is fine as a temporary solution, but it would be great if we can get Wagner to provide us with an updated color system for Material 3 like what he did for the POS project 1qcjzXitBHU7xPnpCOWnNM-fi-162_18857

irfano replied:
I’m not asking for this PR but can’t we use the same colors he used for POS?  I remember Wagner wanted the design system to be consistent across POS and all other parts of the app.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/theme/WooColors.kt:
```diff
@@ -2,63 +2,118 @@ package com.woocommerce.android.ui.compose.theme
+        onSecondary = colorResource(R.color.color_on_secondary),
+        secondaryContainer = colorResource(R.color.color_secondary),
+        onSecondaryContainer = colorResource(R.color.color_on_secondary),
+        tertiary = colorResource(R.color.color_secondary),
+        onTertiary = colorResource(R.color.color_on_secondary),
+        tertiaryContainer = colorResource(R.color.color_secondary),
+        onTertiaryContainer = colorResource(R.color.color_on_secondary),
```
We don't have a tertiary color for now, so I'm applying the secondary one for now.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/theme/WooColors.kt:
```diff
@@ -2,63 +2,118 @@ package com.woocommerce.android.ui.compose.theme
-    val md_theme_light_surface = Color(0xFFFFFFFF)
-    val md_theme_light_onSurface = Color(0xFF000000)
+// Material 2 colors for backward compatibility
+val Material2LightColors
+    @Composable
+    get() = lightColors(
+        primary = colorResource(R.color.color_primary),
+        primaryVariant = colorResource(R.color.color_primary_variant),
+        secondary = colorResource(R.color.color_secondary),
+        secondaryVariant = colorResource(R.color.color_secondary_variant),
+        background = colorResource(R.color.default_window_background),
+        surface = colorResource(R.color.color_surface),
+        error = colorResource(R.color.color_error),
+        onPrimary = colorResource(R.color.color_on_primary),
+        onSecondary = colorResource(R.color.color_on_secondary),
+        onBackground = colorResource(R.color.color_on_background),
+        onSurface = colorResource(R.color.color_on_surface),
+        onError = colorResource(R.color.color_on_error),
+    )
```
This update enables us to maintain a single source of truth for colors, making it easier to update them as needed, and it ensures the consistency of the Compose and View themes.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/theme/Shapes.kt:
```diff
@@ -0,0 +1,19 @@
+import androidx.compose.ui.unit.dp
+import androidx.compose.material.Shapes as Material2Shapes
+
+val Material3Shapes = Shapes(
+    small = RoundedCornerShape(4.dp),
+    medium = RoundedCornerShape(8.dp),
+    large = RoundedCornerShape(8.dp),
+    extraLarge = RoundedCornerShape(8.dp)
+)
+
+val Material2Shapes = Material2Shapes(
+    small = RoundedCornerShape(4.dp),
+    medium = RoundedCornerShape(8.dp),
+    large = RoundedCornerShape(8.dp)
+)
```
I defined the shapes following what we use in our theme today, this will allow to need less overriding for the shapes in code.
This allowed removing the shape override in `WCModalBottomSheet` too.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/theme/WooColors.kt:
```diff
@@ -2,63 +2,118 @@ package com.woocommerce.android.ui.compose.theme
+    get() = lightColorScheme(
+        primary = colorResource(R.color.color_primary),
+        onPrimary = colorResource(R.color.color_on_primary),
+        primaryContainer = colorResource(R.color.color_primary),
```
For now, our components that use the primaryContainer color should use the `primary` color, so this is intended.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/theme/Theme.kt:
```diff
@@ -25,24 +27,39 @@ fun WooTheme(
-    Surface(color = MaterialTheme.colors.background) {
-        content()
+    Surface(color = MaterialTheme.colorScheme.background) {
+        Material2Surface(color = Material2Theme.colors.background) {
+            content()
+        }
```
The background colors are the same.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/theme/WooColors.kt:
```diff
@@ -2,63 +2,118 @@ package com.woocommerce.android.ui.compose.theme
+        inverseSurface = colorResource(R.color.color_on_surface),
+        inversePrimary = colorResource(R.color.color_on_primary),
+        surfaceBright = colorResource(R.color.color_surface),
+        surfaceContainer = colorResource(R.color.color_surface_elevated),
```
We don't have any other colors for now, so we use the same color.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/theme/Typography.kt:
```diff
@@ -93,3 +92,98 @@ val WooTypography = Typography(
+        fontWeight = FontWeight.Normal,
+        fontSize = 24.sp,
+        lineHeight = 24.sp
+    ),
```
`headlineMedium` is equivalent to `h4`. So fontSize should be 34, `lineHeight` should be 36.

hichamboushaba replied:
Even though the official mapping [here](https://developer.android.com/static/develop/ui/compose/images/migration-typography-update.png) states that h4 maps to `headlineMedium`, I don't think it makes sense for us, `headlineMedium` by default has `28px`, while `h4` has `34px`, we use the default `h4` size in our app, so IMO we should map it to `headlineLarge` which by default has `32px`, and keep `headlineMedium` to its default `28px`, please let me know what you think about this.

I don't think we use `headlineMedium` anywhere in the app anyway, so this doesn't seem to have any impact for now.

irfano replied:
Good point! It's better to set them to their default sizes, since they’re not being used in the app and won’t cause any changes.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/compose/theme/Typography.kt:
```diff
@@ -93,3 +92,98 @@ val WooTypography = Typography(
+        fontSize = 20.sp,
+        lineHeight = 24.sp,
+        letterSpacing = 0.5.sp
+    ),
```
`headlineMedium` is equivalent to `h5`. So `fontSize` should be 24 `fontWeight` should be Normal.

I used the table in this [page](https://developer.android.com/develop/ui/compose/designsystems/material2-material3#typography). I wonder how did you determine these sizes but I believe we should set these same as Material 2 by using the table.

Currently I found some font changes and there can be more.

|Before|After|
|-|-|
|<img width="1344" height="2992" alt="before-refund" src="https://github.com/user-attachments/assets/4163012b-ffca-4e6c-a7b0-09af3be42ec8" />|<img width="1344" height="2992" alt="after-refund" src="https://github.com/user-attachments/assets/1d5eb4e0-9fe2-4b91-b58a-68d6827e4d18" />|
|<img width="1344" height="2992" alt="before-light-notice" src="https://github.com/user-attachments/assets/a8f1ad80-5eb7-4d0d-b201-cd42cb8d4c93" />|<img width="1344" height="2992" alt="after-light-notice" src="https://github.com/user-attachments/assets/2368f93d-a222-49a0-ab9d-f4e761818e26" />|
|<img width="1344" height="2992" alt="before-notice" src="https://github.com/user-attachments/assets/af127692-646c-4bfe-8287-8218a73672e8" />|<img width="1344" height="2992" alt="after-notice" src="https://github.com/user-attachments/assets/a981f8e2-ace5-4d7a-a1dc-a8b502d1872e" />|
|<img width="1344" height="2992" alt="before-customs" src="https://github.com/user-attachments/assets/1c9771bd-d57b-4362-ae4b-c24ebd36fc04" />|<img width="1344" height="2992" alt="after-customs" src="https://github.com/user-attachments/assets/5d629ac7-620b-4606-a74a-6f501cb90623" />|

hichamboushaba replied:
Can you please check the fonts now after the last commits? I updated the mappings, and for the Hazmat type, I set a specific text style, because the default style after applying the Material Theme [is](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/material3/material3/src/commonMain/kotlin/androidx/compose/material3/MaterialTheme.kt;l=108?q=MaterialThem) `bodyLarge`.

**Important**: Some small changes are still expected, when I said in the PR description that this PR shouldn't cause any visual changes, it was because I didn't know we were already using Material3 text styles in our screens. Given we were already using them, then updating the styles to match ours will in some cases cause some changes, but these are wanted changes, as they bring more consistency to the app. One noticeable example here is `titleLarge`, the default size for this is `22px`, but in our styles we use `20px`, so whenever it's used, the text will be slightly smaller.

-------

For the color changes, I'll discuss them in a separate comment.

irfano reviewed (changes requested):
This is a great addition, thanks for handling it, Hicham.
I noticed some changes in fonts and colors, and I believe this PR shouldn’t introduce any visual changes. I’ve explained the font differences I observed. Also, please check the background color on the notices.

|Before|After|
|-|-|
|<img width="1344" height="515" alt="before-light-notice" src="https://github.com/user-attachments/assets/54b30330-3d4a-4467-864c-ec42cdb3e08b" />|<img width="1344" height="506" alt="after-light-notice" src="https://github.com/user-attachments/assets/abec5eb9-ee9a-4f07-b033-8f83cd64a276" />|
|<img width="1344" height="604" alt="before-notice" src="https://github.com/user-attachments/assets/6972e824-4fa0-4ccd-a37b-f11ac792a70a" />|<img width="1344" height="512" alt="after-notice" src="https://github.com/user-attachments/assets/f53543ec-7e1b-4cc2-ac61-f4e953c57fe3" />||

hichamboushaba commented:
Thanks @irfano for the great review and for sharing the official typography mapping, even though I visited the above link, I didn't notice the differences because I used AI to generate the typography mapping, thinking it will be one of the tasks where it will do a good job, but it wasn't the case 🤦.
I'll fix the discrepancies, except for `h4`, let's discuss it.

hichamboushaba committed:
Update the material3 typography to correctly map from the material2 ones

Following https://developer.android.com/develop/ui/compose/designsystems/material2-material3#typography

hichamboushaba committed:
Specify a text style for the hazmat type text

hichamboushaba committed:
Update the shipping labels Snackbar to match the app style

hichamboushaba commented:
For the Snackbar color changes, we can divide the discussion two areas:

#### The Snackbar background
The new color is what we should be using, the old one was dark on dark, which didn't provide enough contrast. The old color was used because as we didn't provide a specific theme definition, the default theme was light by default.
The new color is also closer to what we use in other screens where Material 2 is used, for example here:
<img width="360" alt="Screenshot_20250730_103212" src="https://github.com/user-attachments/assets/5f9a140e-1b27-4e83-999f-a15e40e9404f" />

#### The action button color
I updated it in addition to other updates to make the Snackbar more closer to the default material look in 1678e7a, with these updates, it looks now like this:

<img width="360"  alt="Screenshot_20250730_103352" src="https://github.com/user-attachments/assets/a2b9073a-3d74-403d-9688-e136ce185991" />
<img width="360" alt="Screenshot_20250730_110809" src="https://github.com/user-attachments/assets/fd977dab-7fcb-448e-bbaa-4c15b8e6bd3b" />

Please let me know what you think.

-------
These issues are what I wanted to avoid when I suggested p1740493990545489/1740491444.393149-slack-C6H8C3G23 to not use Material 3 components before theming the app.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/wooshippinglabels/components/ShippingLabelsSnackbar.kt:
```diff
@@ -22,13 +22,21 @@ import com.woocommerce.android.R
 fun ShippingLabelsSnackbar(visuals: ShippingLabelsSnackbarVisuals, action: () -> Unit) {
-    Surface(
-        color = SnackbarDefaults.color,
+    Snackbar(
```
As I see it, we used to use a `Surface` as a custom snackbar, but you replaced it with a real `Snackbar`. We’re using Snackbars in other parts of the shipping labels flow, so this change improves consistency. I especially appreciate that.

irfano reviewed (approved):
Thanks for addressing the feedback. I agree with all the points in your comments. I checked all the screens where we use Material 3, including the WooPOS screens, and couldn’t find any other differences.
Great job! We can now safely use Material 3 moving forward. LGTM! 👍🏻
